### ISSUE TITLE
GCC vectorizes less than Intel

### ISSUE THREAD

**Issue by dholladay00** [kokkos/kokkos#871]
It's not clear if this is the proper forum for this. I have an application built on Kokkos and it appears to vectorize fairly well with intel 17.0.1. 

When running gcc 6.1.0, it does not appear to vectorize as well, specifically with math library calls. If I am reading the intel vectorization advisor correctly, the gcc version is not using any vector math library calls (vector exp, log, etc.). 

I have a attached the profiles which can be read with the `kp_reader` executable that is built as a part of the `simple_kernel_timer` found in `kokkos-tools`. The `*_intel` profiles can be as much as twice as fast as the `*_gcc` profiles. 

Anyone have experience getting gcc to include vector math library calls? 

This can be moved offline if needed as it is not an issue with Kokkos itself. 

[profiles.zip](https://github.com/kokkos/kokkos/files/1056203/profiles.zip)

**Comment by dholladay00** [comment]
Worst case results:

```
intel:
                                                                                                zbar         0.07969                    1         0.07969  99.060  77.598
                                                                        Kokkos::View::initialization         0.00073                  149         0.00000   0.906   0.709
                                                                              Setting atomic weights         0.00002                    1         0.00002   0.030   0.023
                                                                      Calculating Block Matrix Sizes         0.00000                    1         0.00000   0.005   0.004

-------------------------------------------------------------------------
Summary:

Total Execution Time (incl. Kokkos + Non-Kokkos:                    0.10270 seconds
Total Time in Kokkos kernels:                                       0.08045 seconds
   -> Time outside Kokkos kernels:                                  0.02225 seconds
   -> Percentage in Kokkos kernels:                                   78.34 %
Total Calls to Kokkos Kernels:                                          152

-------------------------------------------------------------------------

gcc:
                                                                                                zbar         0.17612                    1         0.17612  98.534  93.091
                                                                        Kokkos::View::initialization         0.00257                  149         0.00002   1.438   1.359
                                                                              Setting atomic weights         0.00004                    1         0.00004   0.021   0.020
                                                                      Calculating Block Matrix Sizes         0.00001                    1         0.00001   0.008   0.007

-------------------------------------------------------------------------
Summary:

Total Execution Time (incl. Kokkos + Non-Kokkos:                    0.18919 seconds
Total Time in Kokkos kernels:                                       0.17874 seconds
   -> Time outside Kokkos kernels:                                  0.01045 seconds
   -> Percentage in Kokkos kernels:                                   94.48 %
Total Calls to Kokkos Kernels:                                          152

```

**Comment by ibaned** [comment]
We've been a bit too busy to get in-depth feedback on this, but offhand the comments I've heard are basically:
1. Its not surprising that GCC vectorizes worse than Intel. You may not be able to get them to the same level.
2. When profiling, it is strongly recommended to do more than 1 second of work. At 0.1 or 0.2 seconds, there are many "noise" and overhead factors that will pollute the accuracy of your results.

**Comment by nmhamster** [comment]
@dholladay00 - what flags are you supplying for GCC? One difference between the compilers is how aggressively they permit floating point vectorization with respect to different answers. Historically GCC will want to be provided with `-ffast-math` to utilize vectorized math routines, in the default configuration (e.g. `-O3`), Intel will go ahead and permit optimization. Can you supply your flags and we might be able to provide more advice. I agree with @ibaned that longer benchmarking is a good idea.

**Comment by dholladay00** [comment]
I can do that, but the assembly shows that non-vector versions of transcendental functions (`callq <exp>`, etc.) are being called. 

The issue appears to be an absence of the vector math library `libmvec` on our machines, at least for the versions of gcc that I am using. I am working on requesting it being installed on the hpc machines, but the request needs to be made by someone significantly above my pay grade.

Yes @nmhamster! I just found out about the `-ffast-math` issue yesterday, thanks!

I looked into using `-mveclibabi=svml`, but I don't think that it the way to go since it relies upon intel compilers being installed.

We can probably close the issue for now and I can reopen if the `libmvec` thing doesn't work.
